(PUP-9337) Update mode check to treat FullControl as insync for windows SYSTEM user

McdonaldSeanp opened the pull request:
This commit updates the mode calculations in puppet to treat "FullControl" as insync for the SYSTEM user in the following specific scenario:

1. the file resource does _not_ manage the group
2. the group set for the file is the "SYSTEM" (this is set automatically in windows when puppet runs as the SYSTEM user and creates a file)

This scenario occurs when a file resource does not manage the group on windows and the puppet run that creates the file is run as SYSTEM. Puppet runs will run as SYSTEM when puppet runs as a service, or during invocations from pxp-agent.

For more details on the underlying problems, see https://tickets.puppetlabs.com/browse/PUP-9337

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

puppetcla commented:
CLA signed by all contributors.

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

Iristyle commented on lib/puppet/util/windows/security.rb:
```diff
@@ -247,11 +246,9 @@ def get_mode(path)
       # if owner and group the same, then user and group modes are the OR of both
       if sd.owner == sd.group
         mode |= ((mode & S_IRWXG) << 3) | ((mode & S_IRWXU) >> 3)
-        #puts "owner: #{sd.group}, 0x#{ace.mask.to_s(16)}, #{mode.to_s(8)}"
```
These were actually in there for devs to uncomment while debugging... I think @joshcooper originally added them, but I've used them from time to time.

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

glennsarti commented on lib/puppet/provider/file/windows.rb:
```diff
@@ -71,7 +71,9 @@ def mode
     begin
-      set_mode(value.to_i(8), resource[:path])
+      managing_group = resource[:group].nil? ? false : true
+      managing_owner = resource[:owner].nil? ? false : true
```
nit: `.nil?` already returns a boolean so no need to use a ternary here.

`managing_owner = !resource[:owner].nil?`

glennsarti commented on lib/puppet/provider/file/windows.rb:
```diff
@@ -86,6 +88,32 @@ def validate
+  #
+  # On windows systems: files laid down when puppet is running as the SYSTEM user will have
+  # the SYSTEM user set as the "group" on them. In these situations we want puppet to ignore
+  # what the permissions should be as long as the system user has FullControl access.
```
Nit: Probably best to document what return types are expected here. From looking at it, we expect an integer or nil

What does a nil integer mean?

glennsarti commented on lib/puppet/provider/file/windows.rb:
```diff
@@ -86,6 +88,32 @@ def validate
+      else
+        # If the SYSTEM account does _not_ have FullControl in this scenario, we should
+        # force the resource out of sync no matter what.
+        Puppet.debug _("Not managing SYSTEM user and SYSTEM user does not have full control, treating mode as out of sync")
```
The message here is a bit misleading.
`Debug: Not managing SYSTEM user and SYSTEM user does not have full control, treating mode as out of sync`
Which resource?
Is this the owner or group?
I feel the msg needs more context to make it actionable for a user.

Also may need to add a note for Translators to NOT translate the word SYSTEM as that's a literal word

glennsarti commented on lib/puppet/type/file/mode.rb:
```diff
@@ -116,8 +116,16 @@ def dirmask(value)
     def insync?(currentvalue)
+      if provider.respond_to?(:munge_windows_system_group)
+        munged_mode = provider.munge_windows_system_group(currentvalue, @should)
+        if munged_mode.nil?
```
nit: you can avoid the branch here
```
return false if munged_mode.nil?
currentvalue = munged_mode
```

glennsarti commented:
Also this is weird:

ACL Before:
``` powershell
C:\Source\tmp\puppet-pr7400\tmp>powershell "get-acl .\bad.txt | Select Owner, Group,AccessToString | fl"


Owner          : NT AUTHORITY\SYSTEM
Group          : NT AUTHORITY\SYSTEM
AccessToString : Everyone Allow  ReadAttributes, ReadPermissions, Synchronize
                 NT AUTHORITY\SYSTEM Allow  FullControl
```

Manifest:
``` puppet
file { 'C:/source/tmp/puppet-pr7400/tmp/bad.txt':
  ensure => present,
#  owner => 'SYSTEM',
  group => 'S-1-5-18',
  mode => '660'
}
```

Debug Log
`bundle exec puppet apply tmp\test.pp --debug`

```
...
Debug: Loaded state in 0.03 seconds
Debug: Loaded state in 0.04 seconds
Info: Applying configuration version '1551409221'
Debug: An attempt to set mode 660 on item C:/source/tmp/puppet-pr7400/tmp/bad.txt would result in the owner, SYSTEM, to have less than Full Control rights. This attempt has been corrected to Full Control
Warning: Attempting to set SYSTEM user access to less than Full Control rights. This may have unintented consequences for operations on this file
Notice: /Stage[main]/Main/File[C:/source/tmp/puppet-pr7400/tmp/bad.txt]/mode: mode changed '0770' to '0660'
Debug: /Stage[main]/Main/File[C:/source/tmp/puppet-pr7400/tmp/bad.txt]: The container Class[Main] will propagate my refresh event
Debug: Class[Main]: The container Stage[main] will propagate my refresh event
Debug: Finishing transaction 52450020
Debug: Storing state
...
```

First this
> This attempt has been corrected to Full Control

Then 
> Attempting to set SYSTEM user access to less than Full Control rights.

That's weird.  The debug line above says it was changing this to Full Control already, why am I getting a warning about less than Full Control

And then finally
> C:\Source\tmp\puppet-pr7400\tmp>powershell "get-acl .\bad.txt | Select Owner, Group,AccessToString | fl"
>
>
> Owner          : NT AUTHORITY\SYSTEM
> Group          : NT AUTHORITY\SYSTEM
> AccessToString : Everyone Allow  ReadAttributes, ReadPermissions, Synchronize
>                  NT AUTHORITY\SYSTEM Allow  FullControl
The _actual_ permissions put down are Full Control, so the Warning isn't needed?

McdonaldSeanp commented:
@glennsarti I've updated the PR per your comments.

Specifically with the situation you highlighted last: It's doing both the FullControl force and warning because puppet is individually checking owner and group, which are both SYSTEM, but only the group is specifically managed. Since it performs the checks for user and group separately it's identifying group as a managed user and letting the user set SYSTEM to less than FullControl and emitting a warning. Then as it subsequently moves to user it identifies that user is SYSTEM but _not_ managed and is performing the force to FullControl. Since the ACL won't have different entries for the same user we end up with the FullControl setting from the owner.

We could try to remedy this by checking if _both_ owner and group are set, and try to find the logic around these scenarios.

But TBH I would rather not. The additional logic added here works, and trying to work around when both owner and group are SYSTEM (and only one is managed, or neither, or both, etc. etc.) seems unnecessary to me. It might look a little weird to the puppet user, but it seems to me this is an edge case. The logic required to work around the edge case seems too complex for little return.

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

glennsarti commented:
> but it seems to me this is an edge case

Owner and Group are set differently when run as a scheduled tasks e.g. Like we do in VRA/VRO.  This is what triggered the original PUP-9106 and PUP-6729 issues.  It's not as edge-case as you may think.

I'm beginning to feel we need some integration tests here to check round-trip-ability of various different combinations of owner, user, perms etc. to make sure we're not breaking anything intentionally

glennsarti commented:
I've added a commit with integration tests for a bunch of combinations of owner, group, mode etc.

puppetcla commented:
CLA signed by all contributors.

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a,windows2019-64a

McdonaldSeanp commented:
@glennsarti I've updated the PR from our discussion. 

Now when both owner and group are SYSTEM and one of owner/group is managed but the other is not puppet will emit debug messages but still end up with FullControl from the non-managed part. I think is what we want in that situation. It also will not emit a warning anymore, though maybe we should emit a specific warning in this situation instead of debug messages?

I've also fixed up the integration tests you wrote, see the last commit in the PR for details.

glennsarti commented on spec/integration/provider/file/windows_spec.rb:
```diff
@@ -2,158 +2,161 @@
+      context_name += " and a group of '#{testcase[:manifest_group]}' in the manifest" unless testcase[:manifest_group].nil?
+
+      context context_name do
+        is_idempotent = testcase[:is_idempotent].nil? | testcase[:is_idempotent]
```
Whoops, typo. Should be `||`

glennsarti commented on spec/integration/provider/file/windows_spec.rb:
```diff
@@ -2,158 +2,161 @@
 require 'puppet_spec/files'
 
+# For some reason the provider test will not filter out on windows when using the
+# :if => Puppet.features.microsoft_windows? method of filtering the tests.
```
I suspect this may be due to some other tests mocking that call..

McdonaldSeanp commented:
@glennsarti Updated to fix the typo you mentioned. Are there any other concerns with this PR? or do you think it's good to go?

glennsarti commented on lib/puppet/util/windows/security.rb:
```diff
@@ -328,15 +330,35 @@ def set_mode(mode, path, protected = true)
-        owner_allow = FILE::FILE_ALL_ACCESS
+        # If owner and group are both SYSTEM but group is unmanaged the control rights of system will be set to FUllControl by
+        # the unmanaged group, so there is no need for the warning
+        if managing_owner && !(isownergroup && !managing_group)
```
`!(isownergroup && !managing_group)` is the same as `(!isownergroup || managing_group)`  Does that read better?

McdonaldSeanp replied:
Yeah, I think it might read a little better.

glennsarti commented:
I've approved the PR.  Right now it's functional.  Now we're just quibbling over small things.

speedofdark commented on lib/puppet/provider/file/windows.rb:
```diff
@@ -86,6 +86,52 @@ def validate
+          # If the SYSTEM account does _not_ have FullControl in this scenario, we should
+          # force the resource out of sync no matter what.
+          #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
+          Puppet.debug _("%{resource_name}: %{mode_part_type} set to SYSTEM. SYSTEM permissions cannot be set below FUllControl ('7')") % { resource_name: resource[:name], mode_part_type: mode_part['type']}
```
Change FUllControl to FullControl (uppercase U as second character).

speedofdark commented on lib/puppet/util/windows/security.rb:
```diff
@@ -328,15 +330,35 @@ def set_mode(mode, path, protected = true)
-        #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
-        Puppet.debug _("An attempt to set mode %{mode} on item %{path} would result in the owner, SYSTEM, to have less than Full Control rights. This attempt has been corrected to Full Control") % { mode: mode.to_s(8), path: path }
-        owner_allow = FILE::FILE_ALL_ACCESS
+        # If owner and group are both SYSTEM but group is unmanaged the control rights of system will be set to FUllControl by
```
same FUllcontrol change to FullControl, please

speedofdark commented on lib/puppet/util/windows/security.rb:
```diff
@@ -328,15 +330,35 @@ def set_mode(mode, path, protected = true)
+          #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
+          Puppet.debug _("An attempt to set mode %{mode} on item %{path} would result in the owner, SYSTEM, to have less than Full Control rights. This attempt has been corrected to Full Control") % { mode: mode.to_s(8), path: path }
+          owner_allow = FILE::FILE_ALL_ACCESS
+        end
```
I'm seeing a lot of differentiation between saying FullControl vs. Full Control. Do we need to worry about that?

McdonaldSeanp replied:
No, we are talking about the same thing.

speedofdark commented on lib/puppet/util/windows/security.rb:
```diff
@@ -328,15 +330,35 @@ def set_mode(mode, path, protected = true)
-        #TRANSLATORS 'SYSTEM' is a Windows name and should not be translated
-        Puppet.debug _("An attempt to set mode %{mode} on item %{path} would result in the group, SYSTEM, to have less than Full Control rights. This attempt has been corrected to Full Control") % { mode: mode.to_s(8), path: path }
-        group_allow = FILE::FILE_ALL_ACCESS
+        # If owner and group are both SYSTEM but owner is unmanaged the control rights of system will be set to FUllControl by
```
FUllControl vs. FullControl, please.
